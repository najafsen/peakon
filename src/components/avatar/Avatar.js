import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { getNameInitials } from '../../lib/string.lib';
import './Avatar.scss';

export const Avatar = memo(({ src, name }) => {
    const [isLoaded, setIsLoaded] = useState(false);

    return (
        <div className="avatar-component">
            <div data-testid="alt" className={cx('avatar-alt', { 'hidden': isLoaded })}>{getNameInitials(name)}</div>
            <img data-testid="image" className={cx('avatar-image', { 'hidden': !isLoaded })} src={src} alt={name} onLoad={() => setIsLoaded(true)} />
        </div>
    );
});

Avatar.propTypes = {
    name: PropTypes.string,
    src: PropTypes.string,
};