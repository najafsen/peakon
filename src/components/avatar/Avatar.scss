$size: 3rem;
$radius: .375rem;
$bg-color: #7FD0C9;
$color: #FFFFFF;

.avatar-component {
    display: inline-block;
    height: $size;
    width: $size;

    div, img {
        width: 100%;
        height: 100%;
        border-radius: $radius;
        background-color: $bg-color;
        color: $color;
        font-weight: bold;
        line-height: $size;
        text-align: center;

        &.hidden {
            display: none;
        }
    }
}