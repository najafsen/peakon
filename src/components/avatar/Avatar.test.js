import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Avatar } from './Avatar';

describe('Avatar', () => {

    test('match snapshot initial state', () => {
        const { asFragment } = render(<Avatar name="Mohsen Najaflou" src="some/src" />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('initial state should only show correct alt text', () => {
        const { queryByTestId } = render(<Avatar name="Mohsen Najaflou" src="some/src" />);
        expect(queryByTestId('alt')).not.toHaveClass('hidden');
        expect(queryByTestId('alt')).toHaveTextContent('MN');
        expect(queryByTestId('image')).toHaveClass('hidden');
    });


    test('initial state should only show correct alt text with single word name', () => {
        const { queryByTestId } = render(<Avatar name="Mohsen" src="some/src" />);
        expect(queryByTestId('alt')).not.toHaveClass('hidden');
        expect(queryByTestId('alt')).toHaveTextContent('M');
        expect(queryByTestId('image')).toHaveClass('hidden');
    });


    test('initial state should only show "n/a" alt text with no name', () => {
        const { queryByTestId } = render(<Avatar name="" src="some/src" />);
        expect(queryByTestId('alt')).not.toHaveClass('hidden');
        expect(queryByTestId('alt')).toHaveTextContent('n/a');
        expect(queryByTestId('image')).toHaveClass('hidden');
    });


    test('matches snapshot when image is loaded', () => {
        const { asFragment, queryByTestId } = render(<Avatar name="Mohsen Najaflou" src="some/src" />);
        fireEvent.load(queryByTestId('image'));

        expect(asFragment()).toMatchSnapshot();
    });

    test('image loaded state should only show the image', () => {
        const { queryByTestId } = render(<Avatar name="Mohsen Najaflou" src="some/src" />);
        fireEvent.load(queryByTestId('image'));

        expect(queryByTestId('alt')).toHaveClass('hidden');
        expect(queryByTestId('alt')).toHaveTextContent('MN');
        expect(queryByTestId('image')).not.toHaveClass('hidden');
    });
});