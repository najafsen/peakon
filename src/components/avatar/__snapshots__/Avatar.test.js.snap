// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar match snapshot initial state 1`] = `
<DocumentFragment>
  <div
    class="avatar-component"
  >
    <div
      class="avatar-alt"
      data-testid="alt"
    >
      MN
    </div>
    <img
      alt="Mohsen Najaflou"
      class="avatar-image hidden"
      data-testid="image"
      src="some/src"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar matches snapshot when image is loaded 1`] = `
<DocumentFragment>
  <div
    class="avatar-component"
  >
    <div
      class="avatar-alt hidden"
      data-testid="alt"
    >
      MN
    </div>
    <img
      alt="Mohsen Najaflou"
      class="avatar-image"
      data-testid="image"
      src="some/src"
    />
  </div>
</DocumentFragment>
`;
